### DISCUSSION TITLE
Preview 1.10.4_07 with both left and right sides allowed after "jumping in"

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#560] [Announcements]
Hey folks,

In this version you can place nodes at the both left and right sides of a submap after jump in.
It is a major change in the Freeplane logic because now the side of a child depends on which node is the root of the submap.
Please test it and report bugs. As always, Freeplane development needs your reports and feedback.

Cheers,
Dimitry

**Comment by macmarrum** [comment]
Great, thanks a lot, Dimitri :rocket: 
So far looks good. My tests will continue.

I have a few questions when it comes to scripting. For the jumped-in view:
- how to check if a node is visible? `node.pathToRoot.find { it.visible }` return the jumped-out root
- how to find the jumped-in root node? `node.mindMap.root` returns the jumped-out root
- how to find out if a node is on the left side in the jumped-in view? `node.left` informs about the jumped-out position

**Reply by dpolivaev** [reply]
I have not changed the API yet. I think we need new methods because the existing methods are not related to a view but to a map as a whole, and the new root is a view property and not a map property. 

Could you make suggestions about the new API?

**Reply by dpolivaev** [reply]
Actually, in the internal API I added rootNode as a parameter of isLeft() and so on

**Reply by dpolivaev** [reply]
@macmarrum Do you think you could suggest the new API methods? If not, I'll wait until someone else needs them

**Reply by macmarrum** [reply]
@dpolivaev
I believe `NodeRO#isVisible()` should reflect the non-visibility of nodes that are hidden after a jump-in.
I propose to add `MapRO#getVisibleRoot()` to return the jump-in root or the main root.
I propose to change `NodeRO#isLeft()` to inform about the relation to the visible root; and to add `NodeRO#isLeftFor(Node)` to be able to check for left/right in relation to any root.
I propose to add `String NodeRO#getSide()` and `Node#setSide(String)` (`null` treated as `DEFAULT`) and deprecate `Node#setLeft(boolean)`.

**Comment by euu2021** [comment]
Nice!! 

I did a quick test here, and it seems to be exactly what I had in mind. Thanks!

Later, I will stress test the poor thing, to see if I find any bug.

**Comment by euu2021** [comment]
A couple of things that I noted:

I got a broken script: I have [this script](https://sourceforge.net/p/freeplane/discussion/758437/thread/e9200db6c3/?limit=25#832e), that uses parts of [this script](https://docs.freeplane.org/#/scripting/Scripts_collection?id=paste-clipboard), and now there is some problem with isNewChildLeft():

![image](https://user-images.githubusercontent.com/77707706/184517940-615d24dd-a403-49ac-9482-c2c47f4f395c.png)

Should I report it as a bug? Or should I just change my script?

**Reply by dpolivaev** [reply]
> Should I report it as a bug? Or should I just change my script?

The second.

Whenever you use parts of Freeplane internal API like in example https://docs.freeplane.org/#/scripting/Scripts_collection?id=paste-clipboard the scrips might break at any time because they use internal interfaces of Freeplane which might change at any time.

Particularly, accessing property `node.delegate` you access an internal object. The failure It is not a bug, it is just a consequence of the use of the internal parts of the program which changed.

**Reply by dpolivaev** [reply]
You described three issues in a single post. It makes it difficult for me to handle it. I am going to split the post by hand. As a consequence, I shall be mentioned as an author, but it is less bad for me. In the future please split your posts into atomic parts yourself.

**Reply by euu2021** [reply]
Ok, now I'm getting the difference between using the public API and using the internal things.

**Reply by euu2021** [reply]
So, I fixed my script by replacing the internal things by `menuUtils.executeMenuItems(['PasteAction',])`.

For the script in the collection, I added a note and sent a PR.

**Comment by dpolivaev** [comment]
Written by @eu2021: 

# First thing about clones:

Starting with this situation:
![image](https://user-images.githubusercontent.com/77707706/184518031-a6167ce7-3fe9-408c-80cc-ef74719ee7b5.png)

scenario 1: Cloning C

1.  Jump into B and move C to the left side
2. jump out
3. Clone C to D.
4. Jump into D. The cloned C will be on the left.

Now, if, instead of cloning, you pasted C, it wouldn't be on the left, on step 4.

scenario 2: Cloning B

1.  Jump into B and move C to the left side
2. jump out
3. Clone B to D.
4. Jump into this last B (the child of D). C will be on the right.

Now, if, instead of cloning, you pasted B, then C would be on the left, on step 4.

So, does the node keep the side? It depends on who is copied one:
|        | Cloning | Pasting |
|--------|---------|---------|
| Parent | NO      | YES     |
| Child  | YES     | NO      |

I don't know what to think about it. It feels unintuitive that the behavior changes. Does it make sense?
Map used in the example:
[example left.zip](https://github.com/freeplane/freeplane/files/9332496/example.left.zip)

**Reply by dpolivaev** [reply]
Before I make it consistent: what should be the right logic? 

- Should the top clones keep their sides (Scenario 1, cloning)?
- Should the clones inside the subtrees keep their sides(Scenario 2, cloning)?
- Also, should changing a side of a clone change the sides of its other clones?
- Should pasting preserve sides?

**Reply by euu2021** [reply]
Now, about clones and copies:

I have a branch where I organize information following a certain structure (a traditional textbook structure). Inside this branch, there are several kinds of nodes. One of these kinds is something like a "debate"  style. When I want to work the "debates" nodes, I like to distribute the nodes on the two sides of the root (now, with the jump in feature, it's possible to do that without creating a new map). So, when I jump in, this change happens:
![image](https://user-images.githubusercontent.com/77707706/184538853-8daf5556-9198-431f-b05d-4338c9fb06ec.png)

I clone those "debates" nodes on another node, in order to rank them in a unified list. When I do this, I expect that, when jumping in the cloned debate node, it also has that two sides organization of the "original node", because, then, I can jump in to work on them directly from the list. Otherwise, I would need to navigate through the clones (that can be many, because there can be many lists), until I find the one where the structure is correct.

---
For me, the most intuitive would be designing it to feel like this: the submap has a left/right distribution that is specific to the node and it's children. Thus, moving away one of the child, reset that child left/right state. So:

- copying or cloning the child (C, in the example above): it won't inherit the side. I.e., jumping into the new parent will show that child on the right side.
- copying or cloning the node (B, in the example above): it preserves the left/right distribution of its children. Changing the sides in any of the clones reflect on the other clones. This, for "clone with branch".

**Comment by dpolivaev** [comment]
Written by @eu2021:  

### Second thing about clones:
First steps, starting with only the root:

1. create a node A from the root
2. create a node B from the root (that goes naturally to the left)
3. Create a node C into B. Resulting in this:

![image](https://user-images.githubusercontent.com/77707706/184518618-a7dd2118-0f07-49e2-aba5-01127f86fb22.png)

Scenario 1: cloning B

1. clone B into A
2. jump into A. B will be on the left.

I think it can be a bit confusing, because the user did not explicitly assign B to the left of the root. It was born there.

Scenario 2: cloning C

1. Now, clone C into A
2. jump into A. C will be on the right

After looking at the previous, scenario, I would think that C would be on the left. 

So, there can be confusion is this specific situation of nodes from the left being cloned to the right. How to solve it?

**Reply by dpolivaev** [reply]
Written by @eu2021:  Actually, I noticed that I get directly affected by the scenario 2, of the first situation. As I explained on another thread, I clone nodes from one branch into another, in order to create a rank with all nodes of this kind. So, looking at my usecase, I concluded that, for me, it makes more sense if C keeps at the left, because that's how I explicitly assigned it to be, on the "original" node. 

In other words, if I do a clone of the type "clone with branch", I need that the nodes inside the branch keep the idea of "I should be on the left, If my parent is behaving like a root".

**Reply by dpolivaev** [reply]
@eu2021 Can you please share your use case? I can not make any sensible decisions without a complete understanding of the use cases.
All reasoning is solely based on the use cases.

The same is also true for the previous post ("The first thing"). I need to know how the functionality is used in order to shape it in a consistent way.

**Reply by euu2021** [reply]
> Actually, I noticed that I get directly affected by the scenario 2, of the first situation [...]

my comment was about the other situation, so I will explain it better there.

**Comment by dpolivaev** [comment]
The current solution of keeping the child node sides after jumping in makes the logic too complicated and less intuitive.
I am going to change it so that if you jump into a node placed at the left side of the root all its child nodes appear at the right unless the user explicitly moves them to the left.

Any objections?

**Reply by macmarrum** [reply]
I understand the difficulties related to technical implementation. I believe a jump-in view should behave similar to a jump-out view. What I mean by this is: if a node is on the left of root, after it is made a jumped-in root, it should have its children on the left, unless a child was set to right for the jumped-in root before.
A clone node should have its own left/right for jumped-in root, the same as for a main root.

**Reply by dpolivaev** [reply]
It is not a "technical" issue. Following this approach what should happen if a node is moved from left to the right on a view with either the real map root or some "intermediate" root? Should all "left" and "right" nodes swap?

On the other side, what should happen to the node side after copy-and-paste, copy-and-clone and inside pasted and copied subtrees, particularly if they are inserted on the other side than before?

I think that we should strive for the easiest implementation satisfying given use cases.
The easiest implementation is to start with all nodes on the right side after each jump-in and to allow the user to move all nodes to the left manually. Any other logic seems to be much harder to define.

**Reply by macmarrum** [reply]
The logic is easy to define, in my opinion: each node remembers a side per root node; for children of a jump-in root, if not already known from a previous jump-in, the side reflects where the child was displayed before jump-in.
Because level-2 children of a jump-in root, which is level 0, follow level-1 children's side, side needs to be recorded only for level-1 children.

**Reply by euu2021** [reply]
To make it more clear what is the point here:

Starting with this situation:
![image](https://user-images.githubusercontent.com/77707706/184534336-1755baf9-fb2d-4e60-9e48-8a097c925632.png)
 Jumping into A, B will be on the right.

Now, moving A to the left of the root:
![image](https://user-images.githubusercontent.com/77707706/184534394-d471954e-7836-46a3-9e9a-a0c85af11f88.png)

Jumping into A, B will be on the left.
This is the current behavior, since JumpIn was introduced.
Now, Dimitry proposes to change it, so, in this last example (A on the left), jumping into A will have B on the right.

---

From my experience:
When I use the left side of the Root, I use it because seeing the nodes growing to the left is a cognitive cue to immediately know "where" I'm in the map. So, when I wrote my thesis, I placed the research extracts to the left, and my text to the right, so it was cognitively comfortable to jump between the two parts. Currently, I put my inboxes to the left, and projects to the right, again, to quickly feel where I am. 

Before the JumpIn feature, I used a lot the "filter selected node", to "focus" on a part of the map. It worked well, but there is the problem of nodes being added to the filtered selection, if they are added outside of it. So, I started using the JumpIn, for that purpose. 

The behavior proposed would make me go back to using "filter selected node", because with that I can see the original directions of growth of the nodes. But, don't take it as an argument against the proposal. Indeed, this period of using the JumpIn was an experience to see how it performs against "filter selected node", for that purpose of "focusing" on a node. And, for me, this last one proved to be superior, because of the possibility of showing the ancestors, which creates breadcrumbs to the Root, as explained [here](https://sourceforge.net/p/freeplane/discussion/758437/thread/284e8eb79d/?limit=25#a60b). The breadcrumbs are useful for a quick change of the scope of the focus. I remember Robert STeiner made [a video showing this](https://sourceforge.net/p/freeplane/bugs/3102/).

So, my conclusion is: I agree with the proposal. FP already has a better way to "focus" on a node, so the JumpIn could be designed as aiming to other uses, more like a fresh new map, independent of the way things appear on the original map.

**Reply by dpolivaev** [reply]
Should the side saved after jump-in be kept or reset when the node is copied and pasted or cloned or moved with or without its parent node?

**Reply by macmarrum** [reply]
> Should the side saved after jump-in be kept or reset when the node is copied and pasted or cloned or moved with or without its parent node?

I envisage it being configurable, like for icons `Preferences…->Behaviour->Copy format with icons`, something like `Keep side-allocation mapping on copy/clone/move`.

**Reply by dpolivaev** [reply]
Which concrete different and specific use cases requiring this kind of configuration do you have in mind?

**Reply by macmarrum** [reply]
I have run through all actions that can be done on a node, which I can think of, and analyzed when `side` should be kept or reset. I conclude that, for me, there is no need for the `Keep side...` configuration after all.

In general, I see that it makes sense for me to **reset side** for most actions, e.g. change parent, insert a new node, paste.

There are some actions which I'd like to see to **keep side (no change)**, e.g. move up/down.

And finally, **inherit**: insert a sibling, drop as a sibling (move, copy).

Here's a detailed map: [test-side.zip](https://github.com/freeplane/freeplane/files/9338291/test-side.zip)

**Reply by euu2021** [reply]
Good analysis 👏

> There are some actions which I'd like to see to keep side (no change), e.g. move up/down.
> 
> And finally, inherit: insert a sibling, drop as a sibling (move, copy).

Let me see if I get it right. What you suggest, in this situation (map and submap views, side by side):
![image](https://user-images.githubusercontent.com/77707706/184694232-3a69cf54-a21a-438d-a073-b78e16f4dab8.png)

 is:

- If, in the main map, I drag and drag B on top of A (i.e., it will stay as a sibling, but now on top), it should inherit from A, so it should move to the right in the submap, resulting in this:
![image](https://user-images.githubusercontent.com/77707706/184694799-9950f215-95e5-4f08-b6ab-12cc5703a2c1.png)

- But, if I had moved B using `Edit->Move and sort->Move node (Sibling up)`, then the B node side shouldn't change in the submap, resulting in this:
![image](https://user-images.githubusercontent.com/77707706/184695246-243ac0e4-e169-48be-adbc-85cbd5e9497e.png)

By the way, that's the current behavior.

Another question: is there any part of the current behavior (1.10.4-pre08) do you suggest changing?

**Reply by macmarrum** [reply]
> Let me see if I get it right. What you suggest, in this situation (map and submap views, side by side):

Yes, you got it right.
It's worth to keep in mind that **drag and drop as a sibling** is a broader operation than **move up/down**. **Drop as sibling** can be done from any node; it requests the node to be moved.

> is there any part of the current behavior (1.10.4-pre08) do you suggest changing?

The following actions work differently:
- move left/right: keeps `side` – I'd expect it to reset `side`
- insert parent: keeps `side` for child, inherits `side` from child for the new parent – I'd expect it to reset `side` for both

**Reply by dpolivaev** [reply]
There is a new preview 1.10.4_09, but your latest suggestions are not implemented yet.

**Reply by euu2021** [reply]
(Terminology clarification: I will start calling the root of the submap as `subroot`)

> * move left/right: keeps `side` – I'd expect it to reset `side`


The problem I see in the suggested behavior is this situation (I will call the "archiving situation", because some users use the left side of the root as an "archive" for nodes that are not relevant anymore on the right side):

1. User creates node A to the right of the root
2. User creates a bunch of nodes inside node A
3. User jumps into node A. All the children will be to the right side (default condition)
4. (still in the submap A,) User moves some children to the left, in order to create a distribution that pleases him. So, the right side nodes will have "default condition", and the ones on the left will have "explicitly set condition".
5. User jumps out
6. User moves node A to the left side of the Root. 
7. User jumps into A. What will him find inside A?

What do you think he should find? I think he should find the same distribution that he created when the node was in the right side.

So, it seems that there are 3 possible behaviors, according to what happens to nodes that have an unassigned condition:

1. Closely related to the map: nodes will respect the side that the subroot is, in relation to the Root, when they are created AND if the subroot is later moved to the other side
2. Loosely related to the map: nodes will respect the side that the subroot is, in relation to the Root, when they are created, but NOT if the subroot is later moved to the other side
3. Completely unrelated to the map: default every children to the right side, and also ignore future changes in the position of the subroot

If I understood correctly, the current behavior is 2 and your suggestion is 1. All the solutions have ups and down sides, and that's why I propose solution 3, as it's the most simple and easy to understand. As I said above: 

> FP already has a better way to "focus" on a node, so the JumpIn could be designed as aiming to other uses, more like a fresh new map, independent of the way things appear on the original map.

The 3rd solution would also fulfill [this request](https://sourceforge.net/p/freeplane/discussion/758437/thread/bdf5bc09/), by Ken Hill, which, maybe, is not a very uncommon use case.

**Reply by macmarrum** [reply]
> 4. (still in the submap A,) User moves some children to the left, in order to create a distribution that pleases him. So, the right side nodes will have "default condition", and the ones on the left will have "explicitly set condition".

When jumping in, `side` of each subroot's child, if not already left or right, is assigned to left or right, inherited from the subroot's position in relation to root. Any node on sublevel >= 2 follows sublevel-1 node's side. Therefore current behavior is 1 (with the correction: ~when they are created~ **when jumping in and not already left or right**).

You can use [test-side.zip](https://github.com/freeplane/freeplane/files/9338291/test-side.zip) to see it in action. The map has Map Conditional Styles showing an icon indicating `side` on each node.


>> move left/right: keeps side – I'd expect it to reset side

> The problem I see in the suggested behavior is this situation (I will call the "archiving situation", because some users use the left side of the root as an "archive" for nodes that are not relevant anymore on the right side)

And just for the sake of clarity, my complete suggestion is more complex (pasted from the mindmap):
```text
Move and sort
    left/right
        where parent == root (in current view)
            side: flip left/rigth
        where parent != root (in current view)
            side: reset
```

**Reply by euu2021** [reply]
> When jumping in, `side` of each subroot's child, if not already left or right, is assigned to left or right, inherited from the subroot's position in relation to root.

Indeed. I thought the side was set on creation.

---
For the sake of clarity: with your suggested behavior, what will the user find inside A, in the step 7 above (`User jumps into A. What will him find inside A?`)? With the current behavior (1.10.4_09), the answer is this:

![javaw_9SiMvDMR3S](https://user-images.githubusercontent.com/77707706/184755484-f0535fc3-3ccf-4b23-b1a0-51b195b4d220.gif)

**Reply by dpolivaev** [reply]
@macmarrum 
> The following actions work differently:
> move left/right: keeps side – I'd expect it to reset side
> insert parent: keeps side for child, inherits side from child for the new parent – I'd expect it to reset side for both

Let us rethink it together. Given the use case of the node sides interpreted as "pro arguments" and "contra argument" or any other reason to assign different sides I think the reason remains valid if the nodes are moved upwards to the root, downwards from the root of if a new parent is inserted. Therefore the current implementation keeps assigned sides in such cases.

Do you have any reasons to change it?

**Reply by macmarrum** [reply]
@dpolivaev I stand by my opinion explained in [test-side.zip](https://github.com/freeplane/freeplane/files/9338291/test-side.zip) that moving a node left/right should reset its `side` (unless moving around root; then flip `side`). Also inserting a parent should reset `side` of the new parent and the child – this is explained in [test-side.zip](https://github.com/freeplane/freeplane/files/9338291/test-side.zip) too. If you have questions to the explanations, I'll be happy to answer them.

Having said that, I won't dwell on the subject if you have a strong opinon about this; I am capable of writing scripts that override the logic and set `side` to what I expect.

Anyway, in the latest version built from `freeplane:jumpin`, moving a node to the left/right behaves quite strange. Based on my sample map [test-side.zip](https://github.com/freeplane/freeplane/files/9338291/test-side.zip), when I move `A right-1` left, it's `side` is reset. When I move it again left (so that root is its parent), its `side` is set to `left`, so that the node ends up on the left or root instead of the expected right.
What also strikes me as odd is that when I move `A right-1` right, its `side` is reset, i.e. a different behavior than for moving left (`side` is kept).

**Reply by macmarrum** [reply]
CORRECTION
What also strikes me as odd is that when I move A right-1 right, its `side` is ~reset~ **kept**, i.e. a different behavior than for moving left (`side` is ~kept~ **reset**).

**Reply by dpolivaev** [reply]
I do not have any strong opinions about the logic now, I am asking questions to learn what you think. I have not looked inside test-side.zip yet, I am still going to do it. Thanks for your explanations and for your patience.

**Reply by euu2021** [reply]
> What also strikes me as odd is that when I move A right-1 right, its `side` is ~reset~ **kept**, i.e. a different behavior than for moving left (`side` is ~kept~ **reset**).

Note: there seems to be a refresh delay in the icon. If you interact with the map, the side flips:
![javaw_UNq6Z6TE0F](https://user-images.githubusercontent.com/77707706/185721379-5ed5874f-4211-4f74-ab64-cb09fb375fad.gif)

What I ask is that I can move my debates nodes around and have them keep the disposition that I set (i.e., pro arguments on the right side, and con arguments on the left side). Seeing my disposition flip wouldn't make sense to me.

**Reply by macmarrum** [reply]
> Note: there seems to be a refresh delay in the icon. If you interact with the map, the side flips:

@dpolivaev The Script Filter which changes the icon of `side` uses `node.delegate`. Would this delay be fixed once `NodeRO#getSide()` is introduced?

> What I ask is that I can move my debates nodes around and have them keep the disposition that I set (i.e., pro arguments on the right side, and con arguments on the left side).

@eu2021 Is that for moving left/right or drag-and-drop (changing parents) or both that you want `side` to be kept?
What about inserting a parent or inserting a node? Keep, inherit or reset?

Please keep in mind that dropping a node onto a sibling will inherit `side`. Maybe this could satisfy you, i.e. to have `side` as you want it in your other debate node by dropping onto an existing node of a certain `side`, and moving could reset `side` the way I envisage it?

**Reply by dpolivaev** [reply]
@macmarrum I have just opened test-side.mm. Great analysis. Based on the file name I was not expecting to find it there.
Now after I have seen your reasons I want to agree entirely.

@eu2021 any objections?

**Reply by euu2021** [reply]
Reading again, I see that I wasn't understanding correctly what macmarrum is saying. He is worried about what happens to `A right-1` when `A right-1` is moved; I'm worried about what happens to `A right-1` when `debate-A` is moved.

If this is so, I have no objections.

**Comment by dpolivaev** [comment]
Please check the latest preview 1.10.4_08, I tried to solve all mentioned issues there.

**Reply by euu2021** [reply]
It makes sense to me. Let's see how the other users think about it.

**Comment by dpolivaev** [comment]
In preview 1.10.4_09 jump actions are moved to menu "navigate" and got some fix hot keys.
Also when the action "move to root" is called on a submap it works similar to jump out.

**Comment by dpolivaev** [comment]
I tried to consider all known issues related to node sides in the latest preview 1.10.4_10. It includes also new API methods and your recent pull requests. Please check it, I am thankful for your feedback and looking forward to your next suggestions.

I noticed that the generated API map contains nodes pointing to the old wiki. Do you have an idea what should replace them?

**Reply by macmarrum** [reply]
>  tried to consider all known issues related to node sides in the latest preview 1.10.4_10. It includes also new API methods and your recent pull requests. Please check it,

That's great :star: :star2: :fireworks: 
I'll have a look and report back.

> I am thankful for your feedback and looking forward to your next suggestions.

Cool :+1: 
Likewise, I'm looking forward to collaborating with you on the next thing :smiley: 

> I noticed that the generated API map contains nodes pointing to the old wiki. Do you have an idea what should replace them?

A good catch.

* [Groovy tutorials (Codehaus)](http://groovy.codehaus.org/Beginners+Tutorial) – `http://groovy.codehaus.org/Beginners+Tutorial` is no more. I cannot find anything similar. You mentioned something about a good tutorial during Ensemble Programming Workshop. Was it _Learn Groovy in 5 Minutes_? And maybe we'll have a tutorial of our own (the scripting docs improvements initiative).
* [Example scripts](https://www.freeplane.org/wiki/index.php/Scripting:_Example_scripts) –  `https://www.freeplane.org/wiki/index.php/Scripting:_Example_scripts` could be set to https://docs.freeplane.org/#/scripting/Scripts_collection. We just need to keep in mind to change it when the restructuring happens: https://github.com/freeplane/docs/discussions/76
* [Scripting API changes](https://www.freeplane.org/wiki/index.php/Scripting:_API_Changes) –  `https://www.freeplane.org/wiki/index.php/Scripting:_API_Changes` lives now in the attic: https://docs.freeplane.org/#/attic/old-mediawiki-content/Scripting!_API_Changes, but I'm not sure the information adds any value today, so the link could be removed.

**Reply by euu2021** [reply]
> I tried to consider all known issues related to node sides in the latest preview 1.10.4_10. It includes also new API methods and your recent pull requests. Please check it, I am thankful for your feedback and looking forward to your next suggestions.

Thanks. I will test it.


> * [Groovy tutorials (Codehaus)](http://groovy.codehaus.org/Beginners+Tutorial) – `http://groovy.codehaus.org/Beginners+Tutorial` is no more. I cannot find anything similar. You mentioned something about a good tutorial during Ensemble Programming Workshop. Was it _Learn Groovy in 5 Minutes_? And maybe we'll have a tutorial of our own (the scripting docs improvements initiative).

[This](https://learnxinyminutes.com/docs/groovy/) is the site mentioned by Dimitry.

**Reply by macmarrum** [reply]
Issues spotted so far:
* https://github.com/freeplane/freeplane/issues/579
* https://github.com/freeplane/freeplane/issues/580

**Reply by macmarrum** [reply]
@dpolivaev About API changes, I'm glad Side is accessible now. As for my other suggestions, I can see you decided to go another way.

What troubles me is that I am still unable to establish if a node is visible regardless of whether jump-in view is active or not. Or to find out which node is the visible root.

In other words, I need to know if jump-in was activated (and on which viewRoot) in order to call `NodeRO#.isVisibleOnView(NodeRO)`. Or am my missing something?

Jump-in view is basically a filter that hides the ancestors, including root, so I am struggling to understand why `NodeRO#isVisible()`, which covers filters, doesn't cover jump-in.

**Reply by macmarrum** [reply]
@dpolivaev 
I've been thinking about `NodeRO#getViewSide()`. The name suggests that `side` is associated with a view. It's unclear what view is meant. 
1. If view is meant as "what is displayed", `viewSide` fails to reflect it. Nodes don't have to be displayed based on their `viewSide`. The information about the side where nodes are displayed is delivered by `NodeRO#isLeft()` and `NodeRO#isLeftOnView(NodeRO)`. BTW, the latter has the same issue: little clarity about what is meant by "on view".
2. If view is meant as "jump-in view", `viewSide` delivers more then that, because it covers also the "jump-out view", i.e. for the main root.

Taking these into consideration and keeping in mind that the meaning of `side` is "node's display side when parent is root", I believe the name should rather be `side`, i.e. `NodeRO#getSide()`.
The enum could be named `Side` or `NodeSide`.

As to the meaning of `ifLeftOnView(...)`, I'd put it as "is displayed on the left when the following node is root". I vote for `isLeftFor(NodeRO)` or `isLeftWhenRoot(NodeRO)`.

**Reply by dpolivaev** [reply]
ViewRoot is not a filter but a different feature. If you jump in / out, the filter is kept.
You have got `ControllerRO#getViewRoot()` to be used together with the new methods.
You can use it together with `NodeRO#isDescendantOf(Node)` to find if the node is a part of the current jumped-in view.

**Reply by dpolivaev** [reply]
Generally, API names deserve a separate discussion. There are too many points in this thread, I am lost.

**Reply by macmarrum** [reply]
> ViewRoot is not a filter but a different feature. If you jump in / out, the filter is kept. 

OK. I see your point of view.

> You have got `ControllerRO#getViewRoot()` to be used together with the new methods. You can use it together with `NodeRO#isDescendantOf(Node)` to find if the node is a part of the current jumped-in view.

Thanks. Now I see the new API.

**Reply by macmarrum** [reply]
> Generally, API names deserve a separate discussion. There are too many points in this thread, I am lost.

Discussion created:
* https://github.com/freeplane/freeplane/discussions/588

**Comment by dpolivaev** [comment]
Having more the one topic per thread makes it harder to track and to response to all of them.

**Comment by EdoFro** [comment]
@dpolivaev, @eu2021, @macmarrum,
I instaled today Freeplane v1.10.4-pre11 and tested for the first time the jumping in and out functionality. It works great!!
I couldn't follow the whole discussion (I didn't understand it right), but the resulting feature works superb.

Things I noted that surprise me for good:

- I usually use ctrl+arrow keys to move nodes. They work the same when jumping in. It stays intuitive. great!
- If I change a child node from the jumped-in root node from one side to the other using ctrl+arrow keys, the order of the nodes stays the same ( `parent.getChildPosition(node) `  ). 
- jumped-in clouds works perfect.
- when moving from one side to the other nodes that are part of a summary node, if only some of them move, then the summary doesn't "show" the information correctly, BUT when jumping out it shows correctly.
  - I like it this way: the mindmap information should not be affected if you jump in or out.
  - if you jump in and change some nodes sides, then you jump out and add a summary node to them, if you jump in again the summary node doesn't show correctly, but it doesn't change what you defined when in jumped-out mode. 👍🏼 

I haven't tried with clones yet.

Thanks for your amazing work.

edo

**Reply by EdoFro** [reply]
> * if you jump in and change some nodes sides, then you jump out and add a summary node to them, if you jump in again the summary node doesn't show correctly, but it doesn't change what you defined when in jumped-out mode.

her a little GIF about what I'm trying to say
(sorry about quality of the gif)
![jump-in-out-summary](https://user-images.githubusercontent.com/35700575/186276467-50d8172f-eb30-4f70-849b-51940d93c22a.gif)

**Reply by EdoFro** [reply]
> I haven't tried with clones yet.

I tried now and it works perfect too. I have nothing to add, just great. 👏🏼

**Comment by euu2021** [comment]
Again, about that situation of the sides defined for each node: 

I must say that I got surprised today when I moved this branch to the left side of the root, and nodes 3, 4 and 5 ended all at the left, instead of distributed like they were before:
![javaw_ZCsRiMubRL](https://user-images.githubusercontent.com/77707706/212568212-2e3cf289-be00-4b3e-b78e-6584025c6c0b.gif)

I still need to reread the complex discussion above, but I would like to say that, from an uninformed point of view, the behavior is odd.
